woltka/coverage.py

```diff
@@ -0,0 +1,150 @@
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+"""Functions for calculating coverage of subject sequences.
+
+The "coverage" refers to the ranges of a subject sequence (e.g., genome) that
+are covered by at least one query sequence (e.g., short sequencing read).
+"""
+
+from os import makedirs
+from os.path import join
+from collections import deque
+
+
+def merge_ranges(ranges):
+    """Merge short, fragmental ranges into long, continuous ones.
+
+    Parameters
+    ----------
+    ranges : list of (int, int)
+        Ranges (start, end) to merge.
+
+    Returns
+    -------
+    list of (int, int)
+        Merged ranges.
+    
+    Notes
+    -----
+    Ranges that have overlaps will be merged into one. For example:
+
+    >>> merge_ranges([(1, 3), (2, 4), (6, 8), (7, 9)])
+    [(1, 4), (6, 9)]
+    """
+    res = []
+    res_append = res.append
+    cstart, cend = None, None
+    for start, end in sorted(ranges):
+        if cend is None:
+            # case 1: no active range, start active range.
+            cstart, cend = start, end
+        elif cend >= start - 1:
+            # case 2: active range continues through this range
+            # extend active range
+            cend = max(cend, end)
+        else:
+            # case 3: active range ends before this range begins
+            # write new range out, then start new active range
+            res_append((cstart, cend))
+            cstart, cend = start, end
+    if cend is not None:
+        res_append((cstart, cend))
+    return res
+
+
+def parse_ranges(rmaps, covers, chunk=10000):
+    """Extract range information from read maps.
+
+    Parameters
+    ----------
+    rmaps : dict of dict
+        Read map data structure.
+    covers : dict of list
+        Subject coverage data structure.
+    chunk : int, optional
+        Auto-compress after adding this many new ranges.
+
+    Notes
+    -----
+    The `covers` data structure is a dictionary, where the key is a tuple of
+    (sample, subject), and the value is a list. The first element of the list
+    is an integer indicating how many new ranges have been added since last
+    merge, and the remaining elements are individual ranges.
+
+    Once the range counter reaches the threshold specified by `chunk`, an
+    operation will be automatically triggered to merge small ranges into large,
+    continuous ones, so as to reduce memory space and to improve downstream
+    operations.
+
+    See Also
+    --------
+    merge_ranges
+    """
+    add_cover = covers.setdefault
+    for sample, (_, subque) in rmaps.items():
+        for subjects in subque:
+            for subject, ranges in subjects.items():
+                cover = add_cover((sample, subject), [0])
+                count = cover[0] + len(ranges)
+                if count >= chunk:
+                    cover[:] = [0] + merge_ranges(cover[1:] + ranges)
```

Comment:
I don't understand the [0] here.

Reply:
I believe that this syntax will make a copy of the merge_ranges result list that would not be made by cover[:] = merge_ranges(cover[1:] + ranges), but I can't say for sure that python won't optimize that away